### PR TITLE
drivers/flash/soc_flash_nrf: FTPAN-242 workaround

### LINKED ISSUE #32118: drivers/flash/soc_flash_nrf: nRF anomaly 242 workaround is not implemented
**Describe the bug**
Some of nRF devices are affected by the 242 anomaly:

NVMC operations during POFWARN cause the CPU to hang
###### *For full description pleas referee relevant errata document

**Expected behavior**
workaround implemented

**Impact**
embedded flash operation might  hang the SoC if POFWARN is in use.

### PR DESCRIPTION
**Pull request by nvlsianpu**
- Introduce support for situation when synchronization back-end aborts operation before it is done. synchronization API will transfer operation return code to the driver shim back.

- Disable POFWARN by writing POFCON before a write or erase operation. Do not attempt to write or erase if EVENTS_POFWARN is already asserted.

fixes #32118

### CODE REVIEW

**Comment by nvlsianpu** [comment]
cc @eugmes

**Comment by ioannisg** [comment]
I would like reviews from @anangl and @de-nordic here

File: drivers/flash/soc_flash_nrf.c [hunk 1]
```diff
@@ -488,3 +523,44 @@ static int write(off_t addr, const void *data, size_t len)
...
+
+static int suspend_pofwarn(void)
+{
+	bool enabled;
+	nrf_power_pof_thr_t pof_thr;
+
+	pof_thr = nrf_power_pofcon_get(NRF_POWER, &enabled);
+
+	if (enabled) {
+		nrf_power_pofcon_set(NRF_POWER, false, NRF_POWER_POFTHR_V27);
```
Comment (anangl) [inline drivers/flash/soc_flash_nrf.c:542]: Do you need to change the threshold here? Wouldn't it be sufficient to just disable the warning? I mean:
```
		nrf_power_pofcon_set(NRF_POWER, false, pof_thr);
```
You could eliminate `pofcon_pof_thr` then (and use `nrf_power_pofcon_get()` in `restore_pofwarn()`).

File: drivers/flash/soc_flash_nrf.c [hunk 2]
```diff
@@ -488,3 +523,44 @@ static int write(off_t addr, const void *data, size_t len)
...
+
+		/* This check need to be reworked once POFWARN event will be
+		 * served by zephyr.
+		 */
+		if (nrf_power_event_get_and_clear(NRF_POWER,
+						  NRF_POWER_EVENT_POFWARN)) {
```
Comment (anangl) [inline drivers/flash/soc_flash_nrf.c:548]: Why is the event actually cleared here? Maybe you could use `nrf_power_event_check()` instead?

File: drivers/flash/soc_flash_nrf.c [hunk 3]
```diff
@@ -73,6 +73,18 @@ static struct k_sem sem_lock;
 #define SYNC_UNLOCK()
 #endif
 
+#ifdef CONFIG_NRF52_ANOMALY_242_WORKAROUND
```
Comment (anangl) [inline drivers/flash/soc_flash_nrf.c:76]:
```suggestion
#if NRF52_ERRATA_242_PRESENT
```

File: drivers/flash/soc_flash_nrf.c [hunk 4]
```diff
@@ -488,3 +523,44 @@ static int write(off_t addr, const void *data, size_t len)
...
+	}
+
+	return 0;
+}
+
+static void restore_pofwarn(void)
+{
+	if (pofcon_enabled) {
+		nrf_power_pofcon_set(NRF_POWER, true, pofcon_pof_thr);
+	}
```
Comment (anangl) [inline drivers/flash/soc_flash_nrf.c:564]: `pofcon_enabled = false;` is missing here.

File: drivers/flash/soc_flash_nrf.c [hunk 5]
```diff
@@ -73,6 +73,18 @@ static struct k_sem sem_lock;
 #define SYNC_UNLOCK()
 #endif
 
+#ifdef CONFIG_NRF52_ANOMALY_242_WORKAROUND
+#include <hal/nrf_power.h>
+static int suspend_pofwarn(void);
+static void restore_pofwarn(void);
+
+#define SUSPEND_CONFLICTING_SYSTEMS() suspend_pofwarn()
+#define RESUME_CONFLICTING_SYSTEMS()  restore_pofwarn()
```
Comment (anangl) [inline drivers/flash/soc_flash_nrf.c:82]: Is it worth naming "conflicting systems"? Maybe just `*_POFWARN()` could work?

**Review by anangl — CHANGES_REQUESTED** [review]

**Comment by nvlsianpu** [comment]
@anangl Reworked.

**Review by anangl — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]
Looks OK

**Review by ioannisg — APPROVED** [review]

**Pull request merged**